File: paystack/src/main/java/co/paystack/android/api/model/TransactionApiResponse.java
package co.paystack.android.api.model;

import android.webkit.URLUtil;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;

import java.io.Serializable;

/**
 * 3DS would give a redirect url at which we can conclude payment
 */
public class TransactionApiResponse extends ApiResponse implements Serializable {

    @SerializedName("reference")
    public String reference;

    @SerializedName("trans")
    public String trans;

    @SerializedName("auth")
    public String auth;

    @SerializedName("otpmessage")
    public String otpmessage;

    public static TransactionApiResponse unknownServerResponse() {
        TransactionApiResponse t = new TransactionApiResponse();
        t.status = "0";
        t.message = "Unknown server response";
        return t;
    }

    public static TransactionApiResponse fromJsonString(String jsonString) {
        try {
            return new Gson().fromJson(jsonString, TransactionApiResponse.class);
        } catch (Exception e) {
            return TransactionApiResponse.unknownServerResponse();
        }
    }

    public boolean hasValidReferenceAndTrans() {
        return (reference != null) && (trans != null);
    }

    public boolean hasValidUrl() {
        return otpmessage != null && URLUtil.isValidUrl(otpmessage);
    }

    public boolean hasValidOtpMessage() {
        return otpmessage != null;
    }

    public boolean hasValidAuth() {
        return auth != null;
    }

}

File: paystack/src/main/java/co/paystack/android/api/model/TransactionApiResponse.kt
package co.paystack.android.api.model

import android.webkit.URLUtil
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class TransactionApiResponse(
    @Json(name = "status")
    val status: String? = null,

    @Json(name = "message")
    val message: String? = null,

    @JvmField
    @Json(name = "reference")
    val reference: String? = null,

    @JvmField
    @Json(name = "trans")
    val trans: String? = null,

    @Json(name = "auth")
    val auth: String? = null,

    @Json(name = "otpmessage")
    val otpmessage: String? = null,

    @Json(name = "countryCode")
    val avsCountryCode: String? = null,
) {
    fun hasValidReferenceAndTrans(): Boolean {
        return reference != null && trans != null
    }

    fun hasValidUrl(): Boolean {
        return otpmessage != null && URLUtil.isValidUrl(otpmessage)
    }

    fun hasValidOtpMessage(): Boolean {
        return otpmessage != null
    }

    fun hasValidAuth(): Boolean {
        return auth != null
    }
}